src/eth.rs

```diff
@@ -295,13 +291,9 @@ impl<'d> EthDriver<'d> {
     esp_idf_eth_spi_ethernet_ksz8851snl
 ))]
 impl<'d> EthDriver<'d> {
-    pub fn new_spi<P: spi::Spi + 'd>(
-        spi: impl Peripheral<P = P> + 'd,
+    pub fn new_spi(
+        spi_driver: &spi::SpiDriver<'d>,
```

Comment:
I think we are in a pretty good shape now. The only thing remaining I think which is a bit annoying is that we do not have the flexibility to pass the SPI driver either by reference, or by value (The latter is more convenient when you don't plan to share the SPI bus with multiple devices - see for reference `SpiDeviceDriver`; but of course means you pay the extra verbosity of having the `EthDriver` having a non-lifetime generic now.). Also, folks using `EthDriver` with the RMII and the `openeth` impls will also pay the price (I guess, for them the generic would be resolved as `()`).

Any thoughts?

Reply:
Hmm ye not sure what would be best course of action here. What do you think?

Reply:
I looked into this again and even together with GPT-4 I couldn't find a way :)
Most likely will need to redo whole ETH from ground up to make it play nice

Reply:
I'll also try to look over the weekend at it. Ideally we should merge this before the upcoming release of the esp-idf-* crates.

Reply:
Weekend keeps postponing, but also we have not released either. Sorry for the delay - perhaps *this* weekend. I basically want to see if the `BorrowMut` pattern we use elsewhere can be ergonomic enough here as well. I think it can if we define
```rust
pub struct RMII {}
pub struct OpenEth {}
pub struct SPI<T> where T: BorrowMut<SpiDriver>`
```

and then we parameterize `EthDriver` by `EthDriver<'d, Q>` where Q: is one of `RMII`, `OpenEth` or `SPI<T>`. This might result in a small win (don't remember Rust's method overloading rules exactly) where we can rename back `new_spi`, `new_rmii`, `new_openeth` all to just `new` except that they would be implemented on different parameterizations of `Q`. Might or might not work, have to check.

Reply:
Maybe make common trait `EthDriver` for all of them? Then pass it in `EspEth`. Then have `OpenEthDriver`, `W5500Driver` and `RMIIDriver` that will implement it.

Reply:
This way we might have to duplicate a lot of code across the the three implementations, and I'm not sure how that's fixing the main issue: which is - when using SPI - provide the user the capability to either pass the SPI driver by reference, or move it into the driver, where the latter would be like 99% of the cases (where they don't need to share the SPI bus with something else).

Anyway - my approach actually works. Ahem, type-checks.
I'll be merging your patch shortly, and then committing my approach on top of it.